### ISSUE TITLE
Oneof is not constructable in TypeScript

### ISSUE THREAD

**Issue by StarpTech** [bufbuild/protobuf-es#530]
**Describe the bug**

Try to create a response message like this.

```proto
message AnalyticsViewRow {
  map<string, Value> value = 1;
}

message GetAnalyticsViewResponse {
  Response response = 1;
  AnalyticsViewResult view = 2;
}

message Value {
  // The kind of value.
  oneof kind {
    // Represents a double value.
    double number_value = 1;
    // Represents a string value.
    string string_value = 2;
    // Represents a boolean value.
    bool bool_value = 3;
  }
}
```

If you want to return this as a response in Node.js it only suggests `kind` as input but at the same time it complains that `value` and `case` is missing. I couldn't find a way to create a valid `GetAnalyticsViewResponse` response payload without type issues.

```
const rows: PlainMessage<AnalyticsViewRow>[] = result.map((row) => {
      return {
        value: {
          myField: {
            kind: ".....",
          }
        },
      };
    });
```

```
 Type '{ value: { myField: { kind: string; }; }; }[]' is not assignable to type 'PlainMessage<AnalyticsViewRow>[]'.
  Type '{ value: { myField: { kind: string; }; }; }' is not assignable to type 'PlainMessage<AnalyticsViewRow>'.
    Types of property 'value' are incompatible.
      Type '{ myField: { kind: string; }; }' is not assignable to type '{ [key: string]: PlainMessage<Value>; [key: number]: PlainMessage<Value>; }'.
        Property 'myField' is incompatible with index signature.
          Type '{ kind: string; }' is not assignable to type 'PlainMessage<Value>'.
            Types of property 'kind' are incompatible.
              Type 'string' is not assignable to type '{ value: number; case: "numberValue"; } | { value: string; case: "stringValue"; } | { value: boolean; case: "boolValue"; } | { case: undefined; value?: undefined; }'.
```

**Comment by timostamm** [comment]
For an explanation what is generated for oneof groups, please refer to the documentation [here](https://github.com/bufbuild/protobuf-es/blob/main/docs/generated_code.md#oneof-groups). We have some examples for accessing the type [here](https://github.com/bufbuild/protobuf-es/blob/main/docs/runtime_api.md#accessing-oneof-groups).

In your example, it should probably be something like:

```ts
  const rows: PlainMessage<AnalyticsViewRow>[] = result.map((row) => {
      return {
        value: {
          myField: {
            kind: {case: "numberValue", value: 123},
          }
        },
      };
    });
```

But that is hard to tell without a reproducible example. Closing this, but if you think there is a bug, please feel free to reopen with more information 🙂.

**Comment by StarpTech** [comment]
Hi @timostamm why did you close the issue? It's legit to request for a better example but give me time to react to it. Were you not able to test the example above?

I was able to make it work with your example above but by not following autocompletion and just copy-paste your code. The IntelliSense is broken for such cases. I think this can be fixed by providing better types.

**Comment by timostamm** [comment]
> why did you close the issue?

Because it was filed as a bug report "Oneof is not constructable in TypeScript" without a reproducible example, and tests in this repository indicate that they are constructable.

Regarding autocompletion, it seems to work for `kind`:

<img width="525" alt="1" src="https://github.com/bufbuild/protobuf-es/assets/4289451/2b4a2df5-124d-4b62-96c0-ce9b69b8d2a5">

As well as `case`:

<img width="550" alt="2" src="https://github.com/bufbuild/protobuf-es/assets/4289451/c3db344c-c7a6-44d3-90b8-2f265fbc30c1">

Although TypeScript is unable to provide completion for the discriminator:

<img width="524" alt="3" src="https://github.com/bufbuild/protobuf-es/assets/4289451/af68281c-b29c-4f70-9175-8758641ae409">

The type of value is narrowed down:

<img width="541" alt="4" src="https://github.com/bufbuild/protobuf-es/assets/4289451/620b900e-8d40-406a-9b5c-4d8560113d20">

This seems to work as expected to me. In general, I think that the hard to parse error messages are by far the biggest downside of mapped types.

> I think this can be fixed by providing better types.

What did you have in mind?

**Comment by StarpTech** [comment]
This is my experience:

![image](https://github.com/bufbuild/protobuf-es/assets/1764424/c8060d4e-0aed-423f-ab28-530bdfe9b174)
![image](https://github.com/bufbuild/protobuf-es/assets/1764424/62178bb8-515f-4b01-a165-54a81ca22fdb)


I think the issue is `PlainMessage`. TypeScript has problems inferring the type behind `[key: string | number]`. When I reassign the value to `Record<string, PlainMessage<AnalyticsViewRowValue>>` before I get auto-complete also for the discriminator.

```ts
    const rows: PlainMessage<AnalyticsViewRow>[] = result.map((row) => {
    const viewRow: Record<string, PlainMessage<AnalyticsViewRowValue>> = {
        a: {
          kind: { case: "stringValue", value: "a" },
        },
      };
  })
```
or when I just use `AnalyticsViewRow` without the `PlainMessage` it works too

![image](https://github.com/bufbuild/protobuf-es/assets/1764424/92ff0ea6-d49e-42b2-976c-fca69f8672f5)

What's the recommended way?

TypeScript 5.1.6
